inst/include/Rcpp/sugar/functions/rowSums.h

```diff
@@ -0,0 +1,963 @@
+// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+//
+// rowSums.h: Rcpp R/C++ interface class library -- rowSums, colSums, rowMeans, colMeans 
+//
+// Copyright (C) 2016 Nathan Russell
+//
+// This file is part of Rcpp.
+//
+// Rcpp is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Rcpp is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Rcpp.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef Rcpp__sugar__rowSums_h
+#define Rcpp__sugar__rowSums_h
+
+namespace Rcpp {
+namespace sugar {
+namespace detail {
+
+
+inline bool check_na(double x) {
+    return ISNAN(x);
+}
+
+inline bool check_na(int x) {
+    return x == NA_INTEGER;
+}
+
+inline bool check_na(Rboolean x) {
+    return x == NA_LOGICAL;
+}
+
+inline bool check_na(SEXP x) {
+    return x == NA_STRING;
+}
+
+inline bool check_na(Rcomplex x) {
+    return ISNAN(x.r) || ISNAN(x.i);
+}
+
+
+inline void incr(double& lhs, double rhs) {
```

Comment:
I agree with Kevin, especially in the case where by local idiom there is no
possibility of the pointer being NULL (which BTW would already be true if
you can pass a reference there).

On Sun, Sep 4, 2016 at 9:38 PM, Kevin Ushey <notifications@github.com>
wrote:

> In inst/include/Rcpp/sugar/functions/rowSums.h
> <https://github.com/RcppCore/Rcpp/pull/551#discussion_r77463324>:
>
> > +}
> > +
> > +inline bool check_na(Rboolean x) {
> > +    return x == NA_LOGICAL;
> > +}
> > +
> > +inline bool check_na(SEXP x) {
> > +    return x == NA_STRING;
> > +}
> > +
> > +inline bool check_na(Rcomplex x) {
> > +    return ISNAN(x.r) || ISNAN(x.i);
> > +}
> > +
> > +
> > +inline void incr(double& lhs, double rhs) {
>
> Minor style issue, but I think it's a preferable to receive arguments that
> will be mutated by pointer, rather than by reference -- it makes it more
> obvious at the call site whether a passed-in parameter might be mutated.
> For example,
>
> do_something(x);
>
> versus
>
> do_something(&x);
>
> In the first case, I think the expectation most would have is that
> do_something(x) will not mutate x, whereby it's more obvious that
> mutation might happen with do_something(&x).
>
> The main downside is the fact that pointers can be NULL while references
> can't, but I think the clarity in code makes up for that. I'll let others
> weigh on whether that change is worth making.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RcppCore/Rcpp/pull/551/files/e92b69626da0e6f08f105e01e2f1ff63988da193#r77463324>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXx1FKiNKUquAVcJPoK1VR4pXRlPxMks5qm3KrgaJpZM4J0m43>
> .
>